vim-patch:8.1.{0105,0114,0116,0126,0138,0154,0479,0542,0936}

VVKot opened the pull request:
Problem:    All tab stops are the same.
Solution:   Add the variable tabstop feature. (Christian Brabandt,
            closes vim/vim#2711)
https://github.com/vim/vim/commit/04958cbaf25eea27eceedaa987adfb354ad5f7fd

VVKot commented:
Discussed in #1419. This ports Vim's `vartabstop` option. Notable changes from the original PR:

1. Variable tab stops are stored as longs (to keep it consistent with other settings), converted to ints on access.
2. Change related to side effects of setting incorrect `shiftwidth` (see code below) was not ported, as #6377 removed such side effects.

https://github.com/vim/vim/blob/f8c6a1718007432812184c28495e8d27ee6c0395/src/option.c#L3233-L3240

Should I port all related patches in the same PR? Below is the list of the ones that I found but I might be missing some:

- 8.1.0114
- 8.1.0116
- 8.1.0126
- 8.1.0138
- 8.1.0154
- 8.1.0479
- 8.1.0542
- 8.1.0936

VVKot commented:
@janlazo I noticed that you marked this for 0.5.1, anything I can help with for taking 0.5.0 over the line?

janlazo commented:
@VVKot Port all of the patches that you listed in https://github.com/neovim/neovim/pull/13851#issuecomment-770262933 , including any patches that it depends on within 2 weeks. I plan to update the 0.5.0 milestone and follow the release checklist at https://github.com/neovim/neovim/issues/11023.

VVKot committed:
vim-patch:8.1.0116: display problem with 'vartabstop' and 'linebreak'

Problem:    Display problem with 'vartabstop' and 'linebreak'. (Chauca
            Fuentes)
Solution:   Call tabstop_padding(). (Christian Brabandt, closes vim/vim#3076)
https://github.com/vim/vim/commit/a87b72cc316e065d66dcbcf7ec1cde330adef3a3

VVKot committed:
vim-patch:8.1.0126: various problems with 'vartabstop'

Problem:    Various problems with 'vartabstop'.
Solution:   Fix memory leak.  Fix crash. Add a few more tests. (Christian
            Brabandt, closes vim/vim#3076)
https://github.com/vim/vim/commit/307ac5c68e9e624ab713136d79f35bb73f780d2d

VVKot committed:
vim-patch:8.1.0138: negative value of 'softtabstop' not used correctly

Problem:    Negative value of 'softtabstop' not used correctly.
Solution:   Use get_sts_value(). (Tom Ryder)
https://github.com/vim/vim/commit/33d5ab3795720b7d986f9f17f660ee9e448466e0

VVKot committed:
vim-patch:8.1.0154: crash with "set smarttab shiftwidth=0 softtabstop=-1"

Problem:    Crash with "set smarttab shiftwidth=0 softtabstop=-1".
Solution:   Fall back to using 'tabstop'. (closes vim/vim#3155)
https://github.com/vim/vim/commit/c9fe5ab3b093803b6e8d03358ba16aca6b6f0db1

VVKot committed:
vim-patch:8.1.0479: failure when setting 'varsofttabstop' to end in a comma

Problem:    Failure when setting 'varsofttabstop' to end in a comma. (Ralf
            Schandl)
Solution:   Reject value with trailing command.  Add test for invalid values
            (closes vim/vim#3544)
https://github.com/vim/vim/commit/64f410742f101d7d5ea9e65503e1e0019605eaa5

VVKot committed:
vim-patch:8.1.0936: may leak memory when using 'vartabstop'

Problem:    May leak memory when using 'vartabstop'. (Kuang-che Wu)
Solution:   Fix handling allocated memory for 'vartabstop'. (closes vim/vim#3976)
https://github.com/vim/vim/commit/55c77cf2ea9c15e1ec75d1faf702ec3c9e325271

VVKot commented:
@janlazo mind taking a look when you have a moment?

VVKot commented on src/nvim/buffer.c:
```diff
@@ -1950,15 +1950,8 @@ void free_buf_options(buf_T *buf, int free_p_ff)
-  if (buf->b_p_vsts_array) {
-    xfree(buf->b_p_vsts_array);
-  }
-  buf->b_p_vsts_array = NULL;
```
Ah that's not right, it shouldn't have been deleted. Will force-push without this later today.

VVKot replied:
Done. @janlazo

janlazo commented on runtime/optwin.vim:
```diff
@@ -810,6 +810,12 @@ call <SID>OptionL("ts")
 call append("$", "shiftwidth\tnumber of spaces used for each step of (auto)indent")
 call append("$", "\t(local to buffer)")
 call <SID>OptionL("sw")
+call append("$", "vartabstop\tlist of number of spaces a tab counts for")
```
Do not remove feature checks. Most files in `runtime/` is maintained via upstream patches so avoid diverging from upstream unless it is necessary.

janlazo commented on src/nvim/buffer.c:
```diff
@@ -5361,8 +5372,8 @@ bool bt_terminal(const buf_T *const buf)
 
-// Return true if "buf" is a "nofile", "acwrite" or "terminal" buffer.
-// This means the buffer name is not a file name.
+// Return TRUE if "buf" is a "nofile", "acwrite", "terminal" or "prompt"
```
`bt_nofile()` is a boolean function so it returns true/false, not TRUE/FALSE. Vim uses TRUE/FALSE to support compilers that don't support `bool`. Neovim targets C99 so `bool` is supported.

janlazo commented on src/nvim/buffer.c:
```diff
@@ -5372,7 +5383,8 @@ bool bt_nofile(const buf_T *const buf)
 }
 
-// Return true if "buf" is a "nowrite", "nofile" or "terminal" buffer.
+// Return TRUE if "buf" is a "nowrite", "nofile", "terminal" or "prompt"
```
Same issue here with changing `true` to `TRUE` even though `bt_dontwrite` is a boolean function.

janlazo commented on src/nvim/edit.c:
```diff
@@ -8668,9 +8674,16 @@ static bool ins_tab(void)
 
-  // When nothing special, insert TAB like a normal character
+  // When nothing special, insert TAB like a normal character.
+  bool ts_not_equal_sw = (tabstop_count(curbuf->b_p_vts_array) > 1)
```
This was not in the patch.

VVKot replied:
Yeah, I just got too annoyed formatting this fancy condition. Reverting.

janlazo commented on src/nvim/indent.c:
```diff
@@ -147,35 +183,34 @@ int set_indent(int size, int flags)
         p++;
       }
 
-      // Set initial number of whitespace chars to copy if we are
```
This wasn't removed in the patch.

janlazo commented on src/nvim/ops.c:
```diff
@@ -332,8 +332,9 @@ static void shift_block(oparg_T *oap, int amount)
-  const int p_ts = (int)curbuf->b_p_ts;
+  int p_sw = (int)get_sw_value_indent(curbuf);
+  long *p_vts = curbuf->b_p_vts_array;
+  const long p_ts = curbuf->b_p_ts;
```
This is not in the patch.

VVKot replied:
As mentioned, I moved some variables from int to long to remove compiler warnings. This variable was declared just above as `int` before. I'll remove the ordering change to make it clearer.

janlazo commented on src/nvim/ops.c:
```diff
@@ -3061,14 +3061,17 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
-      coladvance_force(getviscol() + (dir == FORWARD));
+      }
+    } else if (curwin->w_cursor.coladd > 0 || gchar_cursor() == NUL) {
+        coladvance_force(getviscol() + (dir == FORWARD));
```
indent

VVKot replied:
Not positive what you mean here, what would be the correct indent?

janlazo replied:
Neovim codestyle mandates 2-space indent. This has 4 space indent.

VVKot replied:
If we're to automate that - where should it go? Into `make lint`? Seems like something you shouldn't care about during the review.

janlazo replied:
https://github.com/neovim/neovim/blob/master/src/clint.py cannot format so `make lint` cannot not autofix lint errors. https://github.com/neovim/neovim/pull/13289 is WIP for clang-format.

janlazo commented on src/nvim/option.c:
```diff
@@ -1973,6 +1976,11 @@ static void didset_options2(void)
   // Parse default for 'wildmode'.
   check_opt_wim();
+
+  xfree(curbuf->b_p_vsts_array);
```
This comes from a different patch. Mention it in the commit message. You can combine multiple vim patches if it is required to pass CI.

VVKot replied:
This PR contains the commit this comes from, so it should be fine? I changed the PR title to indicate all commits that it contains for clarity.

VVKot replied:
I see what you mean, will change it to reflect the actual commit history.

janlazo commented on src/nvim/option.c:
```diff
@@ -3187,7 +3258,7 @@ did_set_string_option(
     if (varp == &(curwin->w_s->b_p_spl)) {
       char_u fname[200];
-      char_u      *q = curwin->w_s->b_p_spl;
+      char_u *q = curwin->w_s->b_p_spl;
```
This was not in the patch.

janlazo commented on src/nvim/option.c:
```diff
@@ -6905,15 +7040,309 @@ int check_ff_value(char_u *p)
+    return 0;
+  }
+
+  newts = (long *)xmalloc((unsigned)(oldts[0] + 1) * sizeof(long));
```
Does CI fail if `xmalloc` is not casted?

VVKot replied:
I'll try.

janlazo commented on src/nvim/options.lua:
```diff
@@ -2989,6 +2989,22 @@ return {
+      defaults={if_true={vi=""}}
+    },
+    {
+      full_name='vartabstop', abbreviation='vts',
```
`P_RBUF` equivalent is missing here.

janlazo commented on src/nvim/testdir/test_breakindent.vim:
```diff
@@ -12,53 +12,85 @@ source view_util.vim
+func Test_breakindent03()
   " simple breakindent test with showbreak set and briopt including sbr
   call s:test_windows('setl briopt=sbr,min:0 sbr=++')
   let lines=s:screen_lines(line('.'),8)
```
There was space added before,after `=`. Check nearby lines for whitespace changes.

VVKot commented:
@janlazo thanks for the review, will address comments later this week. If that's cutting too close to the 0.5 release - I am happy to wait for 0.5.1 with this PR.

One meta question - should I preserve all 9 commits (i.e. this will be merged without squashing) or should I just make one commit with combined message from all 9 patches?

janlazo commented:
@VVKot Fix the issue in your port of the vim patch and then adjust the commits that follow after it. When reviewing vim-patch PRs, the original vim-patch and the ported version of that vim-patch are compared. You can apply the individual fixes separately and then rebase+squash related commits when you're ready.

I'm going to move this to 0.5.1 release because I haven't reviewed the other commits yet. vartabs feature was harder to review than I expected.

VVKot committed:
vim-patch:8.1.0105: all tab stops are the same

Problem:    All tab stops are the same.
Solution:   Add the variable tabstop feature. (Christian Brabandt,
            closes vim/vim#2711)
https://github.com/vim/vim/commit/04958cbaf25eea27eceedaa987adfb354ad5f7fd

VVKot committed:
vim-patch:8.1.0114: confusing variable name

Problem:    Confusing variable name.
Solution:   Rename new_ts to new_vts_array.  Change zero to NULL.
https://github.com/vim/vim/commit/0119a59ffdfb21cf1c0a56e7ed6105e875150163

VVKot committed:
vim-patch:8.1.0542: shiftwidth() does not take 'vartabstop' into account

Problem:    shiftwidth() does not take 'vartabstop' into account.
Solution:   Use the cursor position or a position explicitly passed.
            Also make >> and << work better with 'vartabstop'. (Christian
            Brabandt)
https://github.com/vim/vim/commit/f951416a8396a54bbbe21de1a8b16716428549f2

VVKot commented:
@janlazo I've addressed the comments, feel free to take a look when you have a moment!

VVKot commented on src/nvim/option.c:
```diff
@@ -7137,7 +7137,7 @@ int tabstop_at(colnr_T col, long ts, long *vts)
   int tab_size = 0;
 
-  if (vts == 0 || vts[0] == 0) {
+  if (vts == NULL || vts[0] == 0) {
```
This check was not changed in the patch, only the other two were. I changed that one too for consistency, should/shouldn't I do that?

janlazo replied:
Sure. Check `NULL`, not 0, for pointers.

janlazo commented on src/nvim/edit.c:
```diff
@@ -8715,7 +8736,9 @@ static bool ins_tab(void)
    * When 'expandtab' not set: Replace spaces by TABs where possible.
    */
-  if (!curbuf->b_p_et && (get_sts_value() || (p_sta && ind))) {
+if (!curbuf->b_p_et && (tabstop_count(curbuf->b_p_vsts_array) > 0
```
indent

janlazo commented on src/nvim/indent.c:
```diff
@@ -396,11 +451,16 @@ int get_breakindent_win(win_T *wp, const char_u *line)
-    prev_indent = get_indent_str(line, (int)wp->w_buffer->b_p_ts, wp->w_p_list);
+    prev_vts = wp->w_buffer->b_p_vts_array;
+    prev_indent = get_indent_str_vtab(line,
+                                      (int)wp->w_buffer->b_p_ts,
```
2nd arg is `long`, not int. Why cast b_p_ts? Same issue with the other `get_indent_` function calls.

VVKot replied:
I just never fixed that after turning args to long. There were `int`s in the original PR, hence we needed casts here.

janlazo replied:
```c
int get_indent_str_vtab(const char_u *ptr, long ts, long *vts, bool list)
```

VVKot replied:
Yep-yep, I fixed this call already!

janlazo commented on src/nvim/indent.c:
```diff
@@ -82,6 +91,30 @@ int get_indent_str(const char_u *ptr, int ts, int list)
 
+// Count the size (in window cells) of the indent in line "ptr", using
+// variable tabstops.
+// if "list" is TRUE, count only screen size for tabs.
```
`list` is a boolean. Use `bool`, not `int`.

janlazo commented on src/nvim/option.c:
```diff
@@ -6911,15 +7043,309 @@ int check_ff_value(char_u *p)
 
+// Set the integer values corresponding to the string setting of 'vartabstop'.
+// "array" will be set, caller must free it if needed.
+int tabstop_set(char_u *var, long **array)
```
bool

VVKot replied:
Are you referring to the second arg here? It is an actual array, even though I see why comment can make one think otherwise.

janlazo replied:
Return value is true/false.

janlazo commented on src/nvim/option.c:
```diff
@@ -6911,15 +7043,309 @@ int check_ff_value(char_u *p)
+    return false;
+  }
+
+  *array = (long *)xmalloc((unsigned)(valcount + 1) * sizeof(long));
```
Vim has 
```c
    *array = (int *) alloc((unsigned) ((valcount + 1) * sizeof(int)));
```

janlazo replied:
Nevermind. Vim changed this later.

janlazo commented on src/nvim/option.c:
```diff
@@ -6911,15 +7043,309 @@ int check_ff_value(char_u *p)
+
 /// Return the effective shiftwidth value for current buffer, using the
-/// 'tabstop' value when 'shiftwidth' is zero.
+/// 'shiftwidth' value when 'shiftwidth' is zero.
```
Vim did not change this.

janlazo commented on src/nvim/option.c:
```diff
@@ -6911,15 +7043,309 @@ int check_ff_value(char_u *p)
+}
+
 /// Return the effective softtabstop value for the current buffer,
 /// using the effective shiftwidth  value when 'softtabstop' is negative.
```
This is what Vim changed instead of `get_sw_value()`. Remove `effective`  from `effective shiftwidth` to match Vim.

VVKot commented:
@janlazo mind taking another look - and I'll rebase if everything is good?

janlazo commented on runtime/doc/eval.txt:
```diff
@@ -7911,6 +7910,11 @@ shiftwidth()						*shiftwidth()*
 
+		When there is one argument {col} this is used as column number
+		for which to return the 'shiftwidth' value. This matters for the
+		'vartabstop'. If no {col} argument is given, column 1 will be
```
Vim has `This matters for the 'vartabstop' feature.`

Keep `feature` or remove `the`.

janlazo commented on test/functional/ex_cmds/excmd_spec.lua:
```diff
@@ -24,8 +24,6 @@ describe('Ex cmds', function()
       pcall_err(command, ':menu 9999999999999999999999999999999999999999'))
     eq('Vim(bdelete):E939: Positive count required',
       pcall_err(command, ':bdelete 9999999999999999999999999999999999999999'))
-    eq('Vim(retab):E487: Argument must be positive',
```
What's the error now?

VVKot replied:
It does not fail anymore.

janlazo commented on src/nvim/option.c:
```diff
@@ -6917,17 +7049,311 @@ int check_ff_value(char_u *p)
+  }
+
+  *array = (long *)xmalloc((unsigned)(valcount + 1) * sizeof(long));
+  if (*array == NULL) {
```
`xmalloc` does not return NULL.

janlazo reviewed (changes requested):
Almost ready. Minor changes remaining.

If there are any missing vartabs patches, you can continue adding them.

janlazo commented:
Rebase so that only `vim-patch:` commits remain and I'll merge.

janlazo commented:
Thank you for your cooperation and hard work.
